custom_components/rrd/__init__.py

```diff
@@ -123,47 +134,95 @@ def setup(hass, config):
 
-        _last_changed_ts = int(datetime.timestamp(state.last_changed))
-        _state = int(state_helper.state_as_number(state))
+        rrd_filename = hass.config.path(rrd_dir, database[CONF_NAME]) + ".rrd"
 
-        ds_name, _, _ = entities[state.entity_id]
-        entities[state.entity_id] = ds_name, _last_changed_ts, _state
+        while not hassIsShuttingDown:
+            # Wait for the begining of next `step` interval.
+            now = time.time()
+            nextSavingTimestamp = ((now // step) + 1) * step
+            time.sleep(nextSavingTimestamp - now)
+
+            if (hassIsShuttingDown):
+                return
+
+
+            # RRD data source names for store.
+            ds_names = []
+            # RRD data source values for store. Corresponding with `ds_names` variable.
+            ds_values = []
+
+            for data_source in database[CONF_DS]:
+                sensor_id = data_source[CONF_SENSOR]
+                ds_name = data_source[CONF_NAME]
+
+                # Get data value
+                sensor_state = hass.states.get(sensor_id)
+                try:
+                    if sensor_state is None:
+                        _LOGGER.debug(
+                            "[%s] Skipping sensor %s, because value is unknown.", rrd_filename, sensor_id
+                        )
+                        raise Exception("Sensor has no value or not exists.")
+
+                    sensor_value = sensor_state.state
+                    # Convert value to integer, when type is COUNTER or DERIVE.
+                    if (data_source[CONF_CF] in ["COUNTER", "DERIVE"]):
+                        sensor_value = round(float(sensor_value))
+                except:
+                    _LOGGER.debug(
+                        "[%s] sensor %s value will be stored as NaN.", rrd_filename, sensor_id
+                    )
+                    sensor_value = "NaN"
+
+                # Add pais of name+value. Will be used as parameters for data save to rrd file.
+                ds_names.append(ds_name)
+                ds_values.append(str(sensor_value))
+
+                try:
+                    template = ":".join(ds_names)
+                    timestamp = int(time.time())
+                    values_string = ":".join(ds_values)
+
+                    rrdtool.update(
+                        rrd_filename, f"-t{template}", f"{timestamp}:{values_string}"
+                    )
+                    _LOGGER.debug(
+                        "%s data added. ds=%s, values=%s:%s", rrd_filename, template, timestamp, values_string
+                    )
+                except rrdtool.OperationalError as exc:
+                    _LOGGER.error(exc)
+
+
+    # Executed on Home assistant start
+    def start(_):
+        try:
+            for database in conf[CONF_DBS]:
+                # Run each database updating in own thread.
+                th = threading.Thread(target=update, args=[database])
```

Comment:
we need to make use of asyncio, HA has several helpers that be used to run the updates in "parallel"

Reply:
Thanks for review, @dgomes . Honestly, I am NodeJS developer - this was my first code in Python. :) So I expected that maybe something not follows the best practicies. Therefore if you can send me any URL, doc or example with the best way for this solution, it will really helps me. I will follow it and rewrite the code.

Reply:
You can use: https://developers.home-assistant.io/docs/asyncio_working_with_async/#starting-independent-task-from-async

Reply:
Maybe have a look into https://github.com/home-assistant/core/blob/c8d550044121ebda46bed76b3edb1eb92623f58b/homeassistant/components/utility_meter/sensor.py who runs periodically also

Reply:
I just noticed that this component is actually not async... 

At the time I decided not to make it async because all database updates are IO intensive... Still we should not start our own threads.

We can probably use `from homeassistant.helpers.event import track_time_change` and be time event driven

Reply:
I found `hass.loop.call_at()`. Function `async_track_time_change` is just wrapper of the `call_at()`. I am currently working on this refactorization.

Reply:
Refactorized to async calls by `hass.loop.call_at()`.
See commit https://github.com/dgomes/rrd_recorder/pull/3/commits/c3ee4d5e2ebc7b399c3163373e35cb6afd16ad69